upload-api/stores/allocations.js

```diff
@@ -0,0 +1,211 @@
+
+/**
+ * @typedef {import('@web3-storage/upload-api/types').AllocationsStorage} AllocationsStorage
+ * @typedef {import('@web3-storage/upload-api/types').BlobAddInput} BlobAddInput
+ * @typedef {import('@web3-storage/upload-api/types').BlobListItem} BlobListItem
+ * @typedef {import('@web3-storage/upload-api/types').ListOptions} ListOptions
+ * @typedef {import('@web3-storage/upload-api/types').DID} DID
+ */
+
+/**
+ * Abstraction layer to handle operations on Store Table.
+ *
+ * @param {string} region
+ * @param {string} tableName
+ * @param {object} [options]
+ * @param {string} [options.endpoint]
+ * @returns {AllocationsStorage}
+ */
+export const createAllocationsStorage = (region, tableName, options = {}) => {
+  const dynamoDb = new DynamoDBClient({
+    region,
+    endpoint: options.endpoint,
+  })
+
+  return useAllocationsStorage(dynamoDb, tableName)
+}
+
+/**
+ * @param {DynamoDBClient} dynamoDb
+ * @param {string} tableName
+ * @returns {AllocationsStorage}
+ */
+export function useAllocationsStorage(dynamoDb, tableName) {
+  return {
+    /**
+     * Check if the given link CID is bound to the uploader account
+     *
+     * @param {import('@ucanto/interface').DID} space
+     * @param {Uint8Array} blobMultihash
+     * @returns {ReturnType<AllocationsStorage['exists']>}
+     */
+    exists: async (space, blobMultihash) => {
+      const key = getKey(space, blobMultihash)
+      const cmd = new GetItemCommand({
+        TableName: tableName,
+        Key: key,
+        AttributesToGet: ['space'],
+      })
+
+      try {
+        const response = await dynamoDb.send(cmd)
+        return { ok: Boolean(response.Item) }
+      } catch {
+        return { ok: false }
+      }
+    },
+    /**
+     * @param {import('@web3-storage/upload-api').DID} space
+     * @param {Uint8Array} blobMultihash
+     * @returns {ReturnType<AllocationsStorage['get']>}
+     */
+    async get(space, blobMultihash) {
+      const key = getKey(space, blobMultihash)
+      const cmd = new GetItemCommand({
+        TableName: tableName,
+        Key: key,
+      })
+
+      const response = await dynamoDb.send(cmd)
+      if (!response.Item) {
+        return { error: new RecordNotFound() }
+      }
+
+      const raw = unmarshall(response.Item)
+      return {
+        ok: {
+          blob: {
+            digest: base58btc.decode(raw.multihash),
+            size: raw.size
+          },
+          invocation: Link.parse(raw.invocation)
+        }
+      }
+    },
+    /**
+     * Bind a link CID to an account
+     *
+     * @param {BlobAddInput} item
+     * @returns {ReturnType<AllocationsStorage['insert']>}
+     */
+    insert: async ({ space, blob, invocation }) => {
+      const insertedAt = new Date().toISOString()
+      const multihash58btc = base58btc.encode(blob.digest)
+
+      const item = {
+        space,
+        multihash: multihash58btc,
+        size: blob.size,
+        invocation: invocation.toString(),
```

Comment:
I realize this just implements what has be previously added to upload-api, however sadly I'm only recognizing this here. It is not a blocker, but perhaps a good idea to do in a followup if we agree.

I am realizing that this should probably link to the signed receipt as opposed to invocation. In a way all our stores should simply be indexes over receipts because receipts are signed commitments from us, also they do link back to the corresponding invocation.

Rational here is that anyone with access to table could create valid record like this, however if it was linking to signed receipt they would not be able to without signing key.

Reply:
I understand your point and this looks like it make sense. I will create an issue to address this as follow up because I am not sure on how we can do this in practise given this is `blob/add` invocation CID. And its receipt link would not be possible to compute when we run `blob/allocate` given it is also dependent on this receipt itself.

Reply:
https://github.com/web3-storage/specs/issues/122

Reply:
Ah sorry I was not very clear here, I meant that it should be receipt to `blob/allocate` not to receipt of the `blob/add`